cairo/tests/fixtures/runner.py

```diff
@@ -58,39 +59,56 @@ def cairo_run(request, cairo_program, cairo_file, main_path):
     """
 
     def _factory(entrypoint, *args, **kwargs):
-        implicit_args = list(
-            cairo_program.identifiers.get_by_full_name(
-                ScopedName(path=("__main__", entrypoint, "ImplicitArgs"))
-            ).members.keys()
-        )
+        implicit_args = cairo_program.identifiers.get_by_full_name(
+            ScopedName(path=("__main__", entrypoint, "ImplicitArgs"))
+        ).members.items()
+
+        # Split implicit args into builtins and other implicit args
+        _builtins = {
+            k: {
+                "python_type": to_python_type(
+                    resolve_main_path(main_path)(v.cairo_type)
+                ),
+                "cairo_type": v.cairo_type,
+            }
+            for k, v in implicit_args
+            if any(builtin in k.replace("_ptr", "") for builtin in ALL_BUILTINS)
+        }
+
+        _implicit_args = {
+            k: {
+                "python_type": to_python_type(
+                    resolve_main_path(main_path)(v.cairo_type)
+                ),
+                "cairo_type": v.cairo_type,
+            }
+            for k, v in implicit_args
+            if not any(builtin in k.replace("_ptr", "") for builtin in ALL_BUILTINS)
+        }
+
         _args = {
-            k: to_python_type(resolve_main_path(main_path)(v.cairo_type))
+            k: {
+                "python_type": to_python_type(
+                    resolve_main_path(main_path)(v.cairo_type)
+                ),
+                "cairo_type": v.cairo_type,
+            }
             for k, v in cairo_program.identifiers.get_by_full_name(
                 ScopedName(path=("__main__", entrypoint, "Args"))
             ).members.items()
         }
-        return_data = cairo_program.identifiers.get_by_full_name(
+
+        implicit_return_data_types = [v["cairo_type"] for v in _implicit_args.values()]
+        explicit_return_data_type = cairo_program.identifiers.get_by_full_name(
             ScopedName(path=("__main__", entrypoint, "Return"))
-        )
+        ).cairo_type
+        return_data_types = implicit_return_data_types + [explicit_return_data_type]
+
         # Fix builtins runner based on the implicit args since the compiler doesn't find them
         cairo_program.builtins = [
             builtin
-            # This list is extracted from the builtin runners
-            # Builtins have to be declared in this order
-            for builtin in [
-                "output",
-                "pedersen",
-                "range_check",
-                "ecdsa",
-                "bitwise",
-                "ec_op",
-                "keccak",
-                "poseidon",
-                "range_check96",
-                "add_mod",
-                "mul_mod",
-            ]
-            if builtin in {arg.replace("_ptr", "") for arg in implicit_args}
+            for builtin in ALL_BUILTINS
+            if builtin in {arg.replace("_ptr", "") for arg in _builtins.keys()}
```

Comment:
can probably replace with a set intersect here

Reply:
not going to use a set here as it changes the order of the builtins, but the cairo-lang library _requires_ them to be in the same order